File: design-pattern-jvm/app/src/main/java/design/pattern/java/patterns/structural/facade/VideoConverter.java
package design.pattern.java.patterns.structural.facade;

public class VideoConverter {
  public ThirdPartyLibraries.VideoFile convert(
      String filename, String format) {
    ThirdPartyLibraries.VideoFile file = new ThirdPartyLibraries.VideoFile(
        filename);
    String sourceCode = ThirdPartyLibraries.CodecFactory.extract(file);
    ThirdPartyLibraries.CompressionCodec destinationCode;
    if (format.matches("^mp4$")) {
      destinationCode = new ThirdPartyLibraries.MPEG4CompressionCodec();
    } else {
      destinationCode = new ThirdPartyLibraries.OggCompressionCodec();
    }
    String buffer = ThirdPartyLibraries.BitrateReader
        .read(filename, sourceCode);
    String result = ThirdPartyLibraries.BitrateReader
        .buffer(buffer, destinationCode);
    return new ThirdPartyLibraries.VideoFile(new ThirdPartyLibraries.AudioMixer().fix(result));
  }
}

File: design-pattern-jvm/app/src/main/kotlin/design/pattern/kotlin/patterns/structural/facade/VideoConverter.kt
package design.pattern.kotlin.patterns.structural.facade

class VideoConverter {
    fun convert(
        filename: String, format: String
    ): ThirdPartyLibraries.VideoFile {
        val file = ThirdPartyLibraries.VideoFile(
            filename
        )
        val sourceCode = ThirdPartyLibraries.CodecFactory.extract(file)
        val destinationCode: ThirdPartyLibraries.CompressionCodec =
            if (format.matches(Regex("^mp4$"))) {
            ThirdPartyLibraries.MPEG4CompressionCodec()
        } else {
            ThirdPartyLibraries.OggCompressionCodec()
        }
        val buffer = ThirdPartyLibraries.BitrateReader
            .read(filename, sourceCode)
        val result = ThirdPartyLibraries.BitrateReader
            .buffer(buffer, destinationCode)
        return ThirdPartyLibraries.VideoFile(ThirdPartyLibraries.AudioMixer().fix(result))
    }
}